pkg/controllers/state/state.go

```diff
@@ -0,0 +1,241 @@
+// compute topology information.
+type Node struct {
+	Node        *v1.Node
+	requested   v1.ResourceList
+	allocatable v1.ResourceList
+
+	podRequests map[types.NamespacedName]v1.ResourceList
+}
+
+// Requested returns the total amount of resources requested by pods that have been bound to the node.
+func (n Node) Requested() v1.ResourceList {
+	return n.requested
+}
+
+// ForPodsWithAntiAffinity calls the supplied function once for each pod with required anti affinity terms that is
+// currently bound to a node. The pod returned may not be up to date with respect to status, however since the
+// anti-affinity terms can't be modified, they will be correct.
+func (c *Cluster) ForPodsWithAntiAffinity(fn func(p *v1.Pod, n *v1.Node) bool) {
+	c.antiAffinityPods.Range(func(key, value interface{}) bool {
+		pod := value.(*v1.Pod)
+		c.mu.RLock()
+		defer c.mu.RUnlock()
+		nodeName, ok := c.bindings[client.ObjectKeyFromObject(pod)]
+		if !ok {
+			return true
+		}
+		node, ok := c.nodes[nodeName]
+		if !ok {
+			// if we receive the node deletion event before the pod deletion event, this can happen
+			return true
+		}
+		return fn(pod, node.Node)
+	})
+}
+
+// ForEachNode calls the supplied function once per node object that is being tracked. It is not safe to store the
+// state.Node object, it should be only accessed from within the function provided to this method.
+func (c *Cluster) ForEachNode(f func(n *Node) bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	for _, node := range c.nodes {
+		if !f(node) {
+			return
+		}
+	}
+}
+
+func (c *Cluster) newNode(node *v1.Node) *Node {
+	n := &Node{
+		Node:        node,
+		podRequests: map[types.NamespacedName]v1.ResourceList{},
+	}
+	var pods v1.PodList
+	if err := c.kubeClient.List(c.ctx, &pods, client.MatchingFields{"spec.nodeName": node.Name}); err != nil {
+		logging.FromContext(c.ctx).Errorf("listing pods, %c", err)
+	}
+	var requested []v1.ResourceList
+	for i := range pods.Items {
+		requests := resources.RequestsForPods(&pods.Items[i])
+		podKey := client.ObjectKeyFromObject(&pods.Items[i])
+
+		n.podRequests[podKey] = requests
+		c.bindings[podKey] = n.Node.Name
+		requested = append(requested, requests)
+	}
+	n.requested = resources.Merge(requested...)
+	n.allocatable = node.Status.Allocatable
+	return n
+}
+
+func (c *Cluster) handleNodeDeletion(nodeName string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	delete(c.nodes, nodeName)
+}
+
+// handleNodeUpdate is called for every node reconciliation
+func (c *Cluster) handleNodeUpdate(node *v1.Node) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	_, ok := c.nodes[node.Name]
+
+	// are we already tracking this node?
+	if !ok {
+		c.nodes[node.Name] = c.newNode(node)
+	} else {
+		// just update the node object so we can track label changes, etc.
+		c.nodes[node.Name].Node = node
+	}
+}
+
+// handlePodDeletion is called when the pod has been deleted
+func (c *Cluster) handlePodDeletion(podKey types.NamespacedName) {
```

Comment:
`handle` is one of those things that could be a prefix to every function, and doesn't add much expressive value.
```suggestion
func (c *Cluster) deletePod(podKey types.NamespacedName) {
```